[DO NOT MERGE]: PR for testing `libcxx` with musl

StefanJum opened the pull request:
This PR contains all commits necessary for building `app-helloworld-cpp` with Musl.
It should only be used for testing, should **NOT** be merged.

It contains the commits from prs: #9 #10 #14 #17 #18, #21 and #22.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

StefanJum committed:
include: Add location of features.h file for Musl

When using Musl, the `features.h` header is not
placed within the `sys/` directory, so the build fails.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

StefanJum committed:
Makefile.uk: Add support headers for Musl

When building with Musl, we need the headers
located in the `include/support/musl` directory
of `libcxx` origin source code.

Without adding the headers in the include path,
the build fails with "xlocale.h not found".

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

StefanJum committed:
patches: Define strtoll_l and strtoull_l in `locale`

The `strtoull_l` and `strtoll_l` functions are needed
by `locale` and are defined as inlines in `xlocale.h`.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

StefanJum committed:
glue.c: Remove multiple definitions when using musl

Musl provides the `catclose`, `catopen` and `catgets`
functions, so when building `libcxx` with musl the
linking fails.

This commit only defines the needed functions when
`LIBMUSL` is not selected.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

StefanJum committed:
[RFC]: Include the first stddef header file

When using Musl, the build breaks because the
`include_next` cannot find a new `stddef.h`
header file.

This is just a hack to go around it and include
the Musl `stddef.h` file, we should find a
proper sollution to this.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

razvand reviewed (changes requested):
Please move this in the `patches/` subdirectory.

StefanJum commented:
Done @razvand.

mariasfiraiala committed:
patches: Define strtof_l and strtod_l in locale

When compiled for Aarch64, apps using libcxx fail due to strtof_l and
strtod_l not being defined in include/locale.

This commit adds a patch which includes the header that contains
the definitions in include/locale.

It also modifies Makefile.uk to add the file in the build process.

Signed-off-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

StefanJum committed:
include/math.h: Add libcxx math.h header

When building `libcxx` with musl, some
undefined functions errors appear because
the musl math.h header is used.

`cmath` looks for some funtions defined in
an anonymous namespace, but they are defines
as macros within the `musl` headers.
This leads to errors since the macros can
not match the required functions.

The functions are defined in `libcxx`
headers, but they are not processed since the
`musl` headers are prioritised.
We cannot prioritise the `libcxx` headers since
it leads to a lot of issues similar to this one.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>

StefanJum commented:
We can close this since all the commits have been merged.
